onnxruntime/core/framework/utils.cc

```diff
@@ -209,6 +212,30 @@ const OrtMemoryInfo& FindMemoryInfoForValue(const SessionState& session_state,
 static common::Status CalculateStaticCopyInfoForFeed(const SessionState& session_state,
                                                      const std::string& input_name,
                                                      MLValueCopyInfo& copy_info) {
+#ifdef ENABLE_TRAINING
```

Comment:
Just to check: this has the same behavior as before if you use a regular `InferenceSession` in a build with `ENABLE_TRAINING`, right?

(I'd mildly prefer that we only hit this different path when we're explicitly doing a partial execution, but I don't feel strongly about it, as long as the behavior for inference-in-a-training-build is unaffected.)

Reply:
Its not exactly the same because previously it would throw an error right away but now it will try to check for intermediate tensor and if it cannot find it then it will throw the same error but with a slighly different message but the same meaning).

---
In reply to: [613560154](https://github.com/microsoft/onnxruntime/pull/7324#discussion_r613560154) [](ancestors = 613560154)

Reply:
can we minimize code copy by:
  ret = session_state.GetInputNodeInfo(input_name, node_info_vec));
  #ifdef ENABLE_TRAINING
  if (!ret.IsOK()) {
   //the else part...
    return Status::OK();
  } 
  #endif
   ...// the same code path as before

Reply:
As previous discussed, the copy is by design as we want to keep inference path untouched as much as possible. A second PR will try to refactor code to bring together training and inferencing so that it is easier to review. This is something that @mrry requested.